exp/addons/controllers/clusterresourceset_controller.go

```diff
@@ -0,0 +1,407 @@
+				errList = append(errList, err)
+				continue
+			}
+
+			if resourceSecret.Type != addonsv1.ClusterResourceSetSecretType {
+				logger.Error(err, "Unsupported ClusterResourceSet Secret type", "Resource type", resourceSecret.Type)
+				conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.WrongSecretTypeReason, clusterv1.ConditionSeverityWarning, "")
+				errList = append(errList, err)
+				continue
+			}
+
+			// Convert the object to unstructured to compare against our before copy.
+			raw, err := runtime.DefaultUnstructuredConverter.ToUnstructured(resourceSecret.DeepCopyObject())
+			if err != nil {
+				logger.Error(err, "Failed to set ClusterResourceSet as resource owner reference",
+					"Resource type", resourceSecret.Type, "Resource name", resourceSecret.Name)
+				errList = append(errList, err)
+			}
+			unstructuredObj := &unstructured.Unstructured{Object: raw}
+
+			err = r.patchOwnerRefToResource(ctx, clusterResourceSet, unstructuredObj)
+			if err != nil {
+				logger.Error(err, "Failed to patch ClusterResourceSet as resource owner reference",
+					"Resource type", resourceSecret.Type, "Resource name", resourceSecret.Name)
+				errList = append(errList, err)
+			}
+			// Since maps are not ordered, we need to order them to get the same hash at each reconcile.
+			keys := make([]string, 0)
+			for key := range resourceSecret.Data {
+				keys = append(keys, key)
+			}
+			sort.Strings(keys)
+
+			for _, key := range keys {
+				dataList = append(dataList, resourceSecret.Data[key])
+			}
+		default:
+			logger.Error(err, "Unsupported ClusterResourceSet resource kind", "Resource Kind", resource.Kind)
+			conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.UnsupportedResourceTypeReason, clusterv1.ConditionSeverityWarning, "")
+			errList = append(errList, err)
+			continue
+		}
+
+		// apply all values in the key-value pair of the resource to the cluster.
+		// As there can be multiple key-value pairs in a resource, each value may have multiple objects in it.
+		isSuccessful := true
+		for i := range dataList {
+			data := dataList[i]
+
+			if err := apply(ctx, remoteClient, data); err != nil {
+				isSuccessful = false
+				logger.Error(err, "failed to apply ClusterResourceSet resource", "Resource kind", resource.Kind, "Resource name", resource.Name)
+				conditions.MarkFalse(clusterResourceSet, addonsv1.ResourcesAppliedCondition, addonsv1.ApplyFailedReason, clusterv1.ConditionSeverityWarning, err.Error())
+				errList = append(errList, err)
+			}
+		}
+
+		binding = generateResourceStatus(isSuccessful, computeHash(dataList))
+		bindings.Resources[resourceKindName] = binding
+	}
+	if len(errList) > 0 {
+		return kerrors.NewAggregate(errList)
+	}
+
+	conditions.MarkTrue(clusterResourceSet, addonsv1.ResourcesAppliedCondition)
+
+	return nil
+}
+
+// patchOwnerRefToResource patches the  ClusterResourceSet as a OwnerReference to the resource.
```

Comment:
```suggestion
// patchOwnerRefToResource adds the ClusterResourceSet as a OwnerReference to the resource.
```

Maybe say which resource are we talking about here?